Doc/library/typing.rst

```diff
@@ -933,6 +933,84 @@ These can be used as types in annotations using ``[]``, each having a unique syn
 
    .. versionadded:: 3.9
 
+
+.. data:: TypeGuard
+
+   Special typing form used to annotate the return type of a user-defined
+   type guard function.  ``TypeGuard`` only accepts a single type argument.
+   At runtime, functions marked this way should return a boolean.
+
+   ``TypeGuard`` aims to benefit *type narrowing* - a technique used by static
+   type checkers to determine a more precise type of an expression within a
+   program's code flow.  Usually type narrowing is done by analyzing
+   conditional code flow and applying the narrowing to a block of code.  The
+   conditional expression here is sometimes referred to as a "type guard"::
+
+      def func(val: Optional[Union[str, float]]):
+          # Non-"None" type guard
+          if val is not None:
+              # Type of val is narrowed to ``Union[str, float]``
+              # "isinstance" type guard
+              if isinstance(val, str):
+                  # Type of val is narrowed to ``str``
+                  ...
+              else:
+                  # Else, type of val is narrowed to ``float``.
+                  ...
+          else:
+              # Type of val remains Optional[Union[str, float]]
+              ...
+
+   Sometimes, a type guard uses a user-defined checking function instead of
+   ``isinstance`` or ``is None`` checks.  These user-defined type guard
+   functions require ``TypeGuard`` to narrow their input types as the static
+   type checker usually does not have enough information to statically infer
+   them.
```

Comment:
```suggestion
   Sometimes it would be convenient to use a user-defined boolean function
   as a type guard.  Such a function should use ``TypeGuard[...]`` as its return type
   to alert static type checkers to this intention.
```